### PR TITLE
Native Epic vs DFP Epic

### PR DESCRIPTION
**Pull request by Mullefa**
## What does this change?

Adds an AB test to test the null hypothesis that there is no difference in the annualised value per impression of like-for-like Native and DFP Epics for readers who don't use AdBlock.

## What is the value of this and can you measure success?

If the null hypothesis is true, then we take a big step towards using DFP to serve Epics. This has lots of potential advantages, some of which include: templating, AB test functionality, real-time optimisation of which Epic to show.

## Screenshots

|             | Native                                                                                                                 | DFP                                                                                                                 |
|-------------|------------------------------------------------------------------------------------------------------------------------|---------------------------------------------------------------------------------------------------------------------|
| __Desktop__ | native_desktop | dfp_desktop |
| __Mobile__  | native_mobile  | dfp_mobile  |

## Tested in CODE?

No. Tested locally.

## Notes

- the DFP Line Item used to target the `epic` slot is called `AB Test - Native Epic vs DFP Epic`
- this Line Item has one associated Creative - `Epic V1.1`
- this Creative has been created from the Native Template `Epic V1` with parameters such that the fields which are in both the acquisition event and Epic component event have matching values e.g. `componentId` is equal to `gdnwb_copts_memco_epic_native_vs_dfp_dfp`

## TODOs

I need to liaise with Devina and Jesse to resolve the `TODO`s before this pull request is merged.

### CODE REVIEW

File: static/src/javascripts/projects/common/modules/experiments/tests/acquisitions-epic-native-vs-dfp.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+import type { AdBlockEpicTemplate } from 'common/modules/commercial/contributions-utilities';
+import { createSlots } from 'commercial/modules/dfp/create-slots';
+
+// The adblockInUse boolean in detect.js is (necessarily) a promise.
+// This does not interact well with the synchronous canRun() function expected by the AB testing framework.
+// To mitigate against this, create a function which returns false <=>
+// ad blocker is active or it is unresolved whether it is active.
+// Since the test running is predicated on this function returning true,
+// readers in the test will have ad blocker off,
+// but we might not include all readers who have ad blocker off.
```
Comment (joelochlann) [inline static/src/javascripts/projects/common/modules/experiments/tests/acquisitions-epic-native-vs-dfp.js]: Do we have a sense of who will be excluded? Is it a "random" (i.e. unpredictable) race condition or is it something systematic (e.g. anyone who's on their first visit)?

File: static/src/javascripts/projects/common/modules/commercial/contributions-utilities.js [hunk 2]
```diff
@@ -282,6 +288,9 @@ const makeABTestVariant = (
             const render = (templateFn: ?EpicTemplate) =>
                 copyPromise
                     .then((copy: AcquisitionsEpicTemplateCopy) => {
+                        if (this.options.isAdSlot) {
+                            return this.options.adSlotEpicTemplate();
```
Comment (joelochlann) [inline static/src/javascripts/projects/common/modules/commercial/contributions-utilities.js]: Would it be simpler to use the normal `template` function and just have `isAdSlot` set in the test file?

**Review by joelochlann — APPROVED** [review]
A few small things - not blockers

File: static/src/javascripts/projects/common/modules/commercial/contributions-utilities.js (hunk 2 above)
Comment (Mullefa) [reply static/src/javascripts/projects/common/modules/commercial/contributions-utilities.js]: yes! 💯

File: static/src/javascripts/projects/common/modules/experiments/tests/acquisitions-epic-native-vs-dfp.js (hunk 1 above)
Comment (Mullefa) [reply static/src/javascripts/projects/common/modules/experiments/tests/acquisitions-epic-native-vs-dfp.js]: random

Comment (Mullefa) [reply static/src/javascripts/projects/common/modules/experiments/tests/acquisitions-epic-native-vs-dfp.js]: worth noting, when I have been developing locally, `isAdBlockerOff()` has always returned the correct value

**Comment by Jester232** [comment]
False

**Review by joelochlann — APPROVED** [review]
Even more approval than previously 🖖

**Pull request merged**